Protobuf support

ackintosh opened the pull request:
Closes #46 

This PR adds protobuf support to the client, which provides functionality to encode metrics to an instance of the [OpenMetrics protobuf](https://github.com/OpenObservability/OpenMetrics/blob/main/proto/openmetrics_data_model.proto) format.

```rust
        let mut registry = Registry::default();

        // ...

        // Encode metrics to an instance of the OpenMetrics protobuf format.
        let metric_set: openmetrics_data_model::MetricSet = encode(&registry);

        for family in metric_set.metric_families {
            println!(family.name);

            for metric in family.metrics {
                // ...
            }
        }
```

ackintosh commented:
@mxinden Because [#47 was broken](https://github.com/prometheus/client_rust/pull/47#issuecomment-1230968472), I created this PR from #47. 🙇  I cherry-picked the commits and added a commit https://github.com/prometheus/client_rust/pull/83/commits/48745c4b77d582166ddac7a05acc27ff03ff3b85.

Tests pass and work fine locally.

mxinden commented:
Following up on https://github.com/prometheus/client_rust/pull/47#discussion_r955669023, i.e. whether we can use the same derive marco. Would the following work?

``` diff
diff --git a/Cargo.toml b/Cargo.toml
index 7367989..33910b1 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -11,7 +11,7 @@ homepage = "https://github.com/prometheus/client_rust"
 documentation = "https://docs.rs/prometheus-client"
 
 [features]
-protobuf = ["dep:prost", "dep:prost-types", "dep:prost-build", "dep:prometheus-client-derive-proto-encode", "dep:void"]
+protobuf = ["dep:prost", "dep:prost-types", "dep:prost-build", "dep:prometheus-client-derive-proto-encode", "dep:void", "prometheus-client-derive-text-encode/protobuf"]
 
 [workspace]
 members = ["derive-text-encode", "derive-proto-encode"]
diff --git a/derive-text-encode/Cargo.toml b/derive-text-encode/Cargo.toml
index f243c24..9dfc825 100644
--- a/derive-text-encode/Cargo.toml
+++ b/derive-text-encode/Cargo.toml
@@ -1,4 +1,5 @@
 [package]
+# TODO: Rename to prometheus-client-derive-encode
 name = "prometheus-client-derive-text-encode"
 version = "0.3.0"
 authors = ["Max Inden <mail@max-inden.de>"]
@@ -9,6 +10,9 @@ repository = "https://github.com/prometheus/client_rust"
 homepage = "https://github.com/prometheus/client_rust"
 documentation = "https://docs.rs/prometheus-client-derive-text-encode"
 
+[features]
+protobuf = []
+
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
diff --git a/derive-text-encode/src/lib.rs b/derive-text-encode/src/lib.rs
index 186aae4..abac180 100644
--- a/derive-text-encode/src/lib.rs
+++ b/derive-text-encode/src/lib.rs
@@ -7,6 +7,11 @@ use syn::DeriveInput;
 
 #[proc_macro_derive(Encode)]
 pub fn derive_encode(input: TokenStream) -> TokenStream {
+    #[cfg(feature = "protobuf")]
+    {
+        todo!()
+    }
+
     let ast: DeriveInput = syn::parse(input).unwrap();
     let name = &ast.ident;
```

mxinden reviewed (commented):
Apart from the suggestion to merge the two derive crates, this looks good to me. Thanks for bearing with me!

ackintosh commented on src/encoding/text.rs:
```diff
@@ -38,7 +38,7 @@ use std::collections::HashMap;
 use std::ops::Deref;
 
-pub use prometheus_client_derive_text_encode::*;
+pub use prometheus_client_derive_encode::*;
```
Would it be good to move this line to `src/encoding.proto` so that consumers import Encode trait as follows?:

```diff
- use prometheus_client::encoding::text::Encode;
+ use prometheus_client::encoding::Encode;
```

mxinden replied:
That makes sense to me. Good idea.

ackintosh replied:
Done via https://github.com/prometheus/client_rust/pull/83/commits/d992badbe2aaecd29a3de82110d3fb9fade3d24b and bumped up the minor version since that is a breaking change. https://github.com/prometheus/client_rust/pull/83/commits/8fce75651c11705fbbfd31a93f18ea75d4d71c06

ackintosh commented:
- Merged the two derive crates
- Fixed missing docs
- Fixed clippy warnings
- Moved `Encode` trait
- Added a changelog entry and bumped up the minor version

@mxinden Please have another look. 🙏

mxinden commented:
Thanks for the follow-ups. I kicked off CI, which is complaining. Can you take a look?

divagant-martian commented:
most if not all the problems are tackled in https://github.com/ackintosh/client_rust/pull/7

divagant-martian commented on derive-encode/src/lib.rs:
```diff
@@ -70,9 +70,87 @@ pub fn derive_encode(input: TokenStream) -> TokenStream {
+                        quote! {
+                            let mut label = prometheus_client::encoding::proto::Label::default();
+                            label.name = #ident_string.to_string();
+                            label.value = format!("{}", self.#ident);
```
This imposes a `Display` requirement on the fields of a struct. Is this actually needed? if so it would be good to document it somewhere because it's unexpected

ackintosh replied:
Thank you for pointing it out. I'm working on that in https://github.com/ackintosh/client_rust/pull/9.

ackintosh replied:
ackintosh#9 has been merged into this PR.

ackintosh commented:
Ahh, Signed-off-by is required...

ackintosh commented:
@mxinden @divagant-martian 

Removed a `Display` requirement from the derive macro (https://github.com/prometheus/client_rust/pull/83/commits/8baf256ff856443a25f42ccd5db78a07eaed8297) and fixed CI errors. I checked that all checks have passed on my repo. https://github.com/ackintosh/client_rust/pull/9

mxinden reviewed (approved):
Wonderful work. I will cut a release soonish.

Thank you @ackintosh for tackling this large project!

ackintosh commented:
Thank you so much for your help! ✨  @mxinden @divagant-martian
